### DISCUSSION TITLE
Apple M2 Max  |  Python 3.12.1  |  ERROR: Invalid requirement: 'pip install Pillow==9.2.0' (from line 22 of requirements_versions.txt)

### DISCUSSION THREAD

**Discussion by whhkhaled** [lllyasviel/Fooocus#2704] [Q&A] [answered]
My Device: Apple M2 Max
My Python: 3.12.1 

I am trying to install Fooocus on my Mac, and i downloaded the requirements one by one, and i belive now i need to download last one "Pillow==9.2.0" but it's not downloading on my device, and it's showing this error:
ERROR: Invalid requirement: 'pip install Pillow==9.2.0' (from line 22 of requirements_versions.txt)
![image](https://github.com/lllyasviel/Fooocus/assets/60566610/bcbbca86-7a0f-4513-8370-feedb71c6a77)

and when i am trying to open Fooocus, i see this error:
![image](https://github.com/lllyasviel/Fooocus/assets/60566610/17f058c5-1cc7-4781-9cf8-fefa966a8e15)

please help me, and also one more qustion, in every time it's downloading again the 6 GB for Fooocus! why it's not saved on my device?

**Comment by Rogala** [comment]
[Pillow supports these Python versions.](https://pillow.readthedocs.io/en/latest/installation/python-support.html#python-support)

**Comment by whhkhaled** [comment]
[image: image.png]

but the python in my device is 3.12.1
Do you think any Pillow version will work for Fooocus AI? or should only
Pillow 9.2.0 as requested from Fooocus AI?

**Comment by mapluisch** [comment, marked as answer]
Hi,

you should follow the tutorial step-by-step incl. creating the conda environment. I also have a M2 Max and just tested it again, everything works fine in this exact order:

1. clone the repo
2. `conda env create -f environment.yaml` within the cloned directory
3. `conda activate fooocus`
4. `pip install -r requirements_versions.txt`
5. `python entry_with_update.py`

Based on your screenshot you already have conda installed, but you're not in the correct environment.

**Reply by whhkhaled** [reply]
how to clone repo

**Reply by whhkhaled** [reply]
please can you enter my device through AnyDesk

**Reply by whhkhaled** [reply]
![Uploading image.png…]()

**Reply by mashb1t** [reply]
@mapluisch thanks a lot for your support and the detailed answer, much appreciated.

@whhkhaled please do not share remote desktop credentials or similar here as this can be heavily misused.
Fooocus requires python3.10, might also work with 3.11 but definitely not 3.12. Make sure to follow the official installation instructions in the readme.

**Reply by whhkhaled** [reply]
hello, please i dont know how to do all this, can you enter my device and do it please???

**Reply by whhkhaled** [reply]
![Uploading image.png…]()

**Comment by whhkhaled** [comment]
can you please help me through AnyDesk ot Team viewer?

**Comment by mapluisch** [comment]
I can try to guide you through the installation steps here, but won't do a remote session. For your own safety, please don't ask random strangers to use AnyDesk or TeamViewer on your system, you never know their intentions.

1. Open a new Terminal window.
2. Enter `git clone https://github.com/lllyasviel/Fooocus.git`and wait for the download to finish.
3. Enter `cd Fooocus` into the same Terminal window.
4. Enter `conda env create -f environment.yaml`, which creates a new conda environment called "fooocus"
5. Enter `conda activate fooocus` to activate said environment
6. Enter `pip install -r requirements_versions.txt` to install all the necessary requirements with matching versions.
7. Finally, run Fooocus via `python entry_with_update.py` (or `python entry_with_update.py --disable-offload-from-vram`).

Hope that helps!

**Reply by whhkhaled** [reply]
![Uploading image.png…]()

**Reply by whhkhaled** [reply]
![Uploading image.png…]()

**Reply by whhkhaled** [reply]
<img width="1263" alt="image" src="https://github.com/lllyasviel/Fooocus/assets/60566610/372dd3bc-d96e-4d5b-bbf4-2ff8173a3024">
i see this error

**Reply by mapluisch** [reply]
I assume you get this error while running the final `pip` - could you please show me the entire error message?

First, run `clear` within Terminal, then re-run the `pip` command, and copy-paste the entire output as comment (please wrap it in a three-backticks block, i.e. \``` your entire log \```)

**Reply by whhkhaled** [reply]
hello sure

**Reply by whhkhaled** [reply]
(fooocus) khaled@Khaleds-MacBook-Pro Fooocus % pip install -r requirements_versions.txt
Requirement already satisfied: torchsde==0.2.5 in /Users/khaled/miniconda3/envs/fooocus/lib/python3.12/site-packages (from -r requirements_versions.txt (line 1)) (0.2.5)
Requirement already satisfied: einops==0.4.1 in /Users/khaled/miniconda3/envs/fooocus/lib/python3.12/site-packages (from -r requirements_versions.txt (line 2)) (0.4.1)
Collecting transformers==4.30.2 (from -r requirements_versions.txt (line 3))
  Downloading transformers-4.30.2-py3-none-any.whl.metadata (113 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 113.6/113.6 kB 1.7 MB/s eta 0:00:00
Collecting safetensors==0.3.1 (from -r requirements_versions.txt (line 4))
  Downloading safetensors-0.3.1.tar.gz (34 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting accelerate==0.21.0 (from -r requirements_versions.txt (line 5))
  Downloading accelerate-0.21.0-py3-none-any.whl.metadata (17 kB)
Collecting pyyaml==6.0 (from -r requirements_versions.txt (line 6))
  Using cached PyYAML-6.0.tar.gz (124 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [54 lines of output]
      running egg_info
      writing lib/PyYAML.egg-info/PKG-INFO
      writing dependency_links to lib/PyYAML.egg-info/dependency_links.txt
      writing top-level names to lib/PyYAML.egg-info/top_level.txt
      Traceback (most recent call last):
        File "/Users/khaled/miniconda3/envs/fooocus/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/khaled/miniconda3/envs/fooocus/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/khaled/miniconda3/envs/fooocus/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 325, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 295, in _get_build_requires
          self.run_setup()
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 311, in run_setup
          exec(code, locals())
        File "<string>", line 288, in <module>
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 104, in setup
          return distutils.core.setup(**attrs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 185, in setup
          return run_commands(dist)
                 ^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
          dist.run_commands()
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 969, in run_commands
          self.run_command(cmd)
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/dist.py", line 967, in run_command
          super().run_command(command)
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/command/egg_info.py", line 321, in run
          self.find_sources()
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/command/egg_info.py", line 329, in find_sources
          mm.run()
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/command/egg_info.py", line 550, in run
          self.add_defaults()
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/command/egg_info.py", line 588, in add_defaults
          sdist.add_defaults(self)
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/command/sdist.py", line 102, in add_defaults
          super().add_defaults()
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/_distutils/command/sdist.py", line 251, in add_defaults
          self._add_defaults_ext()
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/_distutils/command/sdist.py", line 336, in _add_defaults_ext
          self.filelist.extend(build_ext.get_source_files())
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "<string>", line 204, in get_source_files
        File "/private/var/folders/fg/_zd47j3j2hlc6bdd4q4snbnm0000gn/T/pip-build-env-nsigu8xp/overlay/lib/python3.12/site-packages/setuptools/_distutils/cmd.py", line 107, in __getattr__
          raise AttributeError(attr)
      AttributeError: cython_sources
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
(fooocus) khaled@Khaleds-MacBook-Pro Fooocus %

**Reply by whhkhaled** [reply]
all this error i have it after i did step 6 ( pip install -r requirements_versions.txt )

**Reply by mashb1t** [reply]
you're still using python3.12, make sure to use the python installation of your venv / conda env by either using the absolute path to pip3 or checking again in the docs of conda / venv.

**Reply by whhkhaled** [reply]
how i can do it, i dont know

**Reply by whhkhaled** [reply]
![Uploading image.png…]()

**Reply by mashb1t** [reply]
Please consider being more polite (not so pushy) and upload screenshots in a viewable manner if you actually want people to help you. We're supporting in our spare time and not as part of our jobs, so please respect this fact.

**Reply by whhkhaled** [reply]
sorry for that, i just so nervous and i will lose my job because of this, sorry again.

**Comment by whhkhaled** [comment]
![Uploading image.png…]()

**Comment by whhkhaled** [comment]
<img width="1580" alt="image" src="https://github.com/lllyasviel/Fooocus/assets/60566610/82131b5a-9958-4d85-8df7-71d293d7bd75">

thank you everyone it's working now! the issue was python, 

i want to ask one qustion, now it's downloading a lots of staff, yesterday somehow i open it but it was with error and it's download everything, now i can see it's downloading everything again. so every time it will download like this from the biganing? and if no, how i can make sure it's 100% saved on my device, and it will not download like this in everytime i open Fooocus.

thank you so mush again.

**Reply by mashb1t** [reply]
Fooocus is built in the way that it downloads missing components and models the first time you access something. If you've never used Fooocus before you'll have to download around 10GB before being able to generate your first image.
Each preset has its own models and LoRAs, which is why switching presets causes these missing files to be downloaded.
This is normal.
Happy generating!

**Comment by whhkhaled** [comment]
You are amazing man, all of you

Thank you so mush ♥️♥️♥️ appreciate all of you

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by mashb1t** [comment]
Truly a discussion of extremes and a rollercoaster of emotions. I'm glad everything worked in the end.

TL;DR
- not reading the instructions in readme
- not waiting for image upload to be finished
- shady and abusable AnyDesk / TeamViewer connection offers
- emotional pressure "just so nervous and i will lose my job because of this"
- overwhelming gratitude

Thanks again @Rogala and @mapluisch for your support.

==== LOCKED ====
